### PR TITLE
Adds a tolerance for temperatures slightly out of bounds when interpolating

### LINKED ISSUE #729: Handling temperatures just outside the library
I've run into an issue with the `interpolation` temperature method near 293.6 K.  With the MCNP ENDF71 library, the coldest light water s(alpha, beta) table is at 293.606 K, above our default temperature of 293.600.  So trying to use the MCNP s(a, b) with interpolation will give

```
 ERROR: Nuclear data library does not contain cross sections for c_H_in_H2O at
        temperatures that bound 294 K.
```

The error is technically correct, but it would be nice if we built in some tolerance for these tiny out-of-bounds errors.  Maybe we also want some way to have the `interpolation` method revert back to `nearest` when the temperature of a cell is very close to a temperature in the data.

### PR DESCRIPTION
**Pull request by joshmay1**
This change is targeted at addressing provided temperatures slightly outside of existing cross section temperatures when in interpolation mode. The allowable distance outside of the existing bounds is specified with the `temperature_tolerance` attribute of the settings.

For example, if cross sections are available at 300K and 600K and a temperature of 299K is specified, the current code would not allow this as it is unbounded on the lower end. With this change, as long as the difference from the bound to the temperature is within the `temperature_tolerance`, the loaded cross sections will snap to the bound instead of interpolating. The default value of the `temperature_tolerance` is still 10K as this parameter is also used in the nearest method for temperatures.

This should close #729.

### CODE REVIEW

File: src/cell.cpp
Comment (lewisgross1296) [inline src/cell.cpp:342]: It could be useful to let the user know the closest temperature at which data is available and/or the closest temperature that would not trigger an error. They could probably look it up, but this error message could help get there faster.

Comment (joshmay1) [reply src/cell.cpp]: Hi Lewis! Thanks for the comment. I just pushed printing of the nearest temperature of the bounds within the error message.

File: src/cell.cpp
Comment (lewisgross1296) [inline src/cell.cpp:342]: I like this change. I'm wondering, is `data::temperature_min` the temperature you want to tell them about (which seems okay since it relates to the file we have), or is `data::tempterature_min - settings::temperature_tolerance` technically better since that is the lowest temp that the code will run for. Same for `data::temperature_max` vs `data::temperature_max + settings::temperature_tolerance`.

**Review by lewisgross1296 — APPROVED** [review]
I added some feedback for the changes in response to my earlier comment. Besides that and the documentation changes, which look good, I'm not sure I'm familiar enough with the other source files to give more thorough review. I also do not have write access, so my review sadly will not unblock the PR. Maybe you could request review from someone who has write access?

File: src/cell.cpp
Comment (joshmay1) [reply src/cell.cpp:342]: I like showing the actual bounds of the data. I see the tolerancing as being a subtle way to prevent unnecessary crashes, rather than an intentional use. Also, I'd rather try to steer users into using real data, rather than data off the map.

Also, it's worth noting the min and max temperatures are global min and max, not necessarily the min and max available for the nuclides within the current cell being assigned a temperature. Technically, this could tell you a certain temperature is okay, even though the cell's nuclides don't actually have data at that temperature (we'd probably hit the same segfault described in #1292). This would be a pretty fringe case, so I think the single global min and max is fine.

File: tests/unit_tests/test_temp_interp.py
Comment (paulromano) [inline tests/unit_tests/test_temp_interp.py:150]: Rather than having this function modify the fake cross sections that were already set up, I think we should just create both the fake U235 and thermal scattering data together (in `make_fake_cross_section`) so that there's no need to edit it after the fact. Does that make sense?

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for this improvement @joshmay1! Sorry it took me a while to review this in full. I have a few small comments but overall it looks like it's in really good shape.

File: tests/unit_tests/test_temp_interp.py
Comment (joshmay1) [reply tests/unit_tests/test_temp_interp.py]: Yep, that makes total sense! I originally didn't want to mess with the existing tests, but this shouldn't have an impact.

File: src/thermal.cpp
Comment (joshmay1) [inline src/thermal.cpp:129]: > Thanks for this improvement @joshmay1! Sorry it took me a while to review this in full. I have a few small comments but overall it looks like it's in really good shape.

You're welcome @paulromano and thank you for the review. Sorry for the goofy comparisons, but those should be fixed now. One thing I noticed is that there doesn't seem to be an explicit check if a thermal scattering library has already been pushed back as there is with the nuclide temperatures. Was there a reason for this?

Comment (paulromano) [reply src/thermal.cpp:129]: I don't think so -- that looks like a bug waiting to happen! Do you want to go ahead and fix that on this brach?

**Review by paulromano — APPROVED** [review]
Everything looks good now -- thanks for the updates @joshmay1! I'll wait for your response about updating the check when pushing back temperatures for thermal scattering before I merge this.

**Comment by joshmay1** [comment]
I think it's good to go now @paulromano. It was indeed reading in libraries multiple times for duplicated values, but only when executing the interpolation tolerance I had just added.

**Comment by paulromano** [comment]
Thanks @joshmay1!

**Pull request merged**
